### PR TITLE
no-outline-none

### PR DESCRIPTION
**Pull request by YozhikM**
#1 
/cc @evilebottnawi

### CODE REVIEW

File: src/rules/no-outline-none/index.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    root.walk(node => {
+      let selector = null;
+
+      if (node.type === 'rule') {
+        selector = node.selector;
+      } else if (node.type === 'atrule' && node.name === 'page' && node.params) {
+        selector = node.params;
+      }
```
Comment (evilebottnawi) [inline src/rules/no-outline-none/index.js]: Need add checks on selector https://github.com/stylelint/stylelint/blob/master/lib/rules/selector-attribute-brackets-space-inside/index.js#L30

File: src/rules/no-outline-none/index.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    if (!validOptions || !actual) {
+      return;
+    }
+
+    root.walk(node => {
+      let selector = null;
+
+      if (node.type === 'rule') {
+        selector = node.selector;
+      } else if (node.type === 'atrule' && node.name === 'page' && node.params) {
```
Comment (evilebottnawi) [inline src/rules/no-outline-none/index.js]: css is case insensitivity use `node.name.toLowerCase()`

File: src/rules/no-outline-none/index.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  expected: value => `Unexpected using "outline" property in ${value}`,
+});
+
+function check(node) {
+  if (node.type !== 'rule') {
+    return true;
+  }
+
+  const hasOutlineNone = node.nodes.some(
+    o => o.type === 'decl' && o.prop === 'outline' && ['0', 'none'].indexOf(o.value) >= 0
```
Comment (evilebottnawi) [inline src/rules/no-outline-none/index.js]: Use `o.prop.toLowerCase()` and `o.value.toLowerCase()`

File: src/rules/no-outline-none/index.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+export const messages = utils.ruleMessages(ruleName, {
+  expected: value => `Unexpected using "outline" property in ${value}`,
+});
+
+function check(node) {
+  if (node.type !== 'rule') {
+    return true;
+  }
+
+  const hasOutlineNone = node.nodes.some(
```
Comment (evilebottnawi) [inline src/rules/no-outline-none/index.js]: Can be simplify, just return `node.nodes.some`

File: src/rules/no-outline-none/__tests__/index.js [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+      column: 4,
+    },
+    {
+      code: '.baz { outline: 0; }',
+      message: messages.expected('.baz'),
+      line: 1,
+      column: 4,
+    },
+  ],
+});
```
Comment (evilebottnawi) [inline src/rules/no-outline-none/__tests__/index.js]: Add tests where used non standard syntaxes (example `scss/less`).

File: src/rules/no-outline-none/index.js [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+      if (!selector) {
+        return;
+      }
+
+      const isAccepted = check(node);
+
+      if (!isAccepted) {
+        utils.report({
+          index: node.lastEach,
```
Comment (YozhikM) [inline src/rules/no-outline-none/index.js]: I doubt this LoC

**Comment by evilebottnawi** [comment]
Better use helpers from `stylelint` as here https://github.com/kristerkari/stylelint-scss/blob/master/src/utils/isStandardSelector.js, i think in near future we will publish version with `stylelint.utils.isStandardSelector`, but we can do this in other PR

**Comment by jeddy3** [comment]
Over in https://github.com/stylelint/stylelint/pull/3355#issuecomment-394426248 I mentioned that this type of enforcement can happen at the configuration level using the limiting rules e.g.

```json
{
  "declaration-property-value-blacklist": {
    "outline": ["none", "0"]
  }
}
```

However, I guess there's no harm publishing this rule within a plugin pack. The pack is in your control and is not subject to the stylelint VISION doc. You'll get an explicit message out of it too, rather than the generic `Unexpected value "${value}" for property "${property}"` one that comes with the `declaration-property-value-blacklist` rule.

Alternatively, you could publish a `stylelint-config-a11y`  shared configuration alongside this plugin pack. It could turn on the rules within the pack and also configure the appropriate limiting rules built-in to stylelint to enforce things like the example above. See https://github.com/stylelint/stylelint/issues/3308#issuecomment-388963505.

Each approach has its pros and cons. I'll let you weigh them up and go whichever approach you think is best.

I'm afraid I won't have time for hands-on code reviews here, but I'll certainly make time to answer questions related to this kind of thing. So, don't wait for a review from me. You can take comfort in the fact that plugin packs can iterate more quickly than stylelint itself, e.g. introduce breaking changes more easily, so I'd push forward with whichever approach works for you as you can go back and change it later.

**Comment by YozhikM** [comment]
I need to add a check for the presence of the border at :focus , and also check that the border is not transparent .

**Comment by YozhikM** [comment]
@jeddy3 
I understand your concerns, but this rule is important for the accessibility of interfaces, when working with the keyboard, for example. 

I'll take your advice about the settings. Thanks!

**Comment by YozhikM** [comment]
/сс @evilebottnawi 
I think I'm done.

**Comment by jeddy3** [comment]
> I think I'm done.

Good job!

> I need to add a check for the presence of the border at :focus , and also check that the border is not transparent .

Then adding a dedicated rule to this plugin pack was a good approach :)

As mentioned in https://github.com/stylelint/stylelint/pull/3355#issuecomment-394426248:

> A little later we realised that not everything a team might need could be achieved using just these configurable rules, so we added a plugin system. This allowed users to create rules which could do more than just apply limits to targeted constructs...

i.e. if configuring a built-in rule does the job, then use that. If it doesn't then, writing a plugin is a great approach.

**Pull request merged**